## Multiple Linear Regression

### Model

In [2]:
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
df = pd.read_csv("Advertising.csv")
df = df.iloc[:,1:len(df)]
df.head()

TV  radio  newspaper  sales
0  230.1   37.8       69.2   22.1
1   44.5   39.3       45.1   10.4
2   17.2   45.9       69.3    9.3
3  151.5   41.3       58.5   18.5
4  180.8   10.8       58.4   12.9

In [3]:
X = df.drop('sales', axis=1)
y = df[["sales"]]

In [4]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [5]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [6]:
#Statsmodels using

In [7]:
import statsmodels.api as sm

In [8]:
lm = sm.OLS(y, X) #OLS calling

In [9]:
model = lm.fit()

In [10]:
model.summary() #these Infos are very important

<class 'statsmodels.iolib.summary.Summary'>
"""
                                 OLS Regression Results                                
=======================================================================================
Dep. Variable:                  sales   R-squared (uncentered):                   0.982
Model:                            OLS   Adj. R-squared (uncentered):              0.982
Method:                 Least Squares   F-statistic:                              3566.
Date:                Thu, 23 Nov 2023   Prob (F-statistic):                   2.43e-171
Time:                        15:12:23   Log-Likelihood:                         -423.54
No. Observations:                 200   AIC:                                      853.1
Df Residuals:                     197   BIC:                                      863.0
Df Model:                           3                                                  
Covariance Type:            nonrobust                                                  
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
TV             0.0538      0.001     40.507      0.000       0.051       0.056
radio          0.2222      0.009     23.595      0.000       0.204       0.241
newspaper      0.0168      0.007      2.517      0.013       0.004       0.030
==============================================================================
Omnibus:                        5.982   Durbin-Watson:                   2.038
Prob(Omnibus):                  0.050   Jarque-Bera (JB):                7.039
Skew:                          -0.232   Prob(JB):                       0.0296
Kurtosis:                       3.794   Cond. No.                         12.6
==============================================================================

Notes:
[1] R² is computed without centering (uncentered) since the model does not contain a constant.
[2] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [11]:
##scikit learn using

In [12]:
from sklearn.linear_model import LinearRegression
lm  = LinearRegression()

In [13]:
model = lm.fit(X, y)

In [14]:
model.intercept_

array([2.93888937])

In [15]:
model.coef_

array([[ 0.04576465,  0.18853002, -0.00103749]])

## Predictive

**Sales = 2.94 + TV * 0.04 + radio * 0.19 - newspaper * 0.001**

30 unit TV, 10 unit radio, 40 unit newspaper 

In [16]:
new_data = [[30],[10],[40]]

In [17]:
import pandas as pd
new_data = pd.DataFrame(new_data).T

In [18]:
new_data

0   1   2
0  30  10  40

In [19]:
model.predict(new_data)

/Users/cbarank/anaconda3/lib/python3.10/site-packages/sklearn/base.py:420: UserWarning: X does not have valid feature names, but LinearRegression was fitted with feature names
  warnings.warn(


array([[6.15562918]])

In [20]:
#and now lets check the MSE and RMSE

In [21]:
from sklearn.metrics import mean_squared_error

In [22]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [23]:
model.predict(X)[0:10]

array([[20.52397441],
       [12.33785482],
       [12.30767078],
       [17.59782951],
       [13.18867186],
       [12.47834763],
       [11.72975995],
       [12.12295317],
       [ 3.72734086],
       [12.55084872]])

In [24]:
MSE = mean_squared_error(y, model.predict(X))
MSE

2.784126314510936

In [25]:
import numpy as np
RMSE = np.sqrt(MSE)
RMSE

1.6685701407225697

## Model Tuning

In [26]:
X.head()

TV  radio  newspaper
0  230.1   37.8       69.2
1   44.5   39.3       45.1
2   17.2   45.9       69.3
3  151.5   41.3       58.5
4  180.8   10.8       58.4

In [27]:
y.head()

sales
0   22.1
1   10.4
2    9.3
3   18.5
4   12.9

In [31]:
#split the test sets and train sets
from sklearn.model_selection import train_test_split

In [32]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.20, random_state = 1)

In [33]:
X_train.head()

TV  radio  newspaper
108   13.1    0.4       25.6
107   90.4    0.3       23.2
189   18.7   12.1       23.4
14   204.1   32.9       46.0
56     7.3   28.1       41.4

In [34]:
y_train.head()

sales
108    5.3
107    8.7
189    6.7
14    19.0
56     5.5

In [35]:
X_test.head()

TV  radio  newspaper
58   210.8   49.6       37.7
40   202.5   22.3       31.6
34    95.7    1.4        7.4
102  280.2   10.1       21.4
184  253.8   21.3       30.0

In [36]:
y_test.head()

sales
58    23.8
40    16.6
34     9.5
102   14.8
184   17.6

In [37]:
lm = LinearRegression()
model = lm.fit(X_train, y_train)

In [38]:
#train error
np.sqrt(mean_squared_error(y_train, model.predict(X_train)))

1.736902590147092

In [39]:
#test error
np.sqrt(mean_squared_error(y_test, model.predict(X_test)))

1.4113417558581565

In [40]:
#lets see teh k cross validation of the model

In [41]:
from sklearn.model_selection import cross_val_score

In [42]:
cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error")

array([-1.62375953, -3.81875608, -3.43828142, -2.27748673, -7.25325414,
       -1.88303708, -2.80517715, -3.68594486, -1.12810834, -3.96330989])

but the values are negative and we dont understand anything?
lets turn to positive and get the mean of values:

In [43]:
#CV mean
np.mean(-cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error"))

3.187711520944358

In [44]:
#CV RMS
np.sqrt(np.mean(-cross_val_score(model, X_train, y_train, cv = 10, scoring = "neg_mean_squared_error")))

1.7854163438661466

In [ ]:
#cv rmse
np.sqrt(np.mean(-cross_val_score(model, X, y, cv = 10, scoring = "neg_mean_squared_error")))

We validated the model